src-native/asami/main.clj

```diff
@@ -1,51 +1,105 @@
 (ns ^{:doc "Entry point for CLI"
       :author "Paula Gearon"}
   asami.main
-  (:require [clojure.string :as s]
-            [clojure.edn :as edn]
+  (:require [asami.core :as asami]
+            [asami.graph :as graph]
             [cheshire.core :as json]
+            [clojure.edn :as edn]
+            [clojure.java.io :as io]
             [clojure.pprint :refer [pprint]]
-            [asami.core :as asami]
-            [asami.graph :as graph])
+            [clojure.string :as s])
+  (:import (java.io PushbackReader))
   (:gen-class))
 
-(def reader-opts {:readers (assoc graph/node-reader 'a/r re-pattern)})
+(def eof
+  (reify))
+
+(defn eof? [x]
+  (identical? x eof))
+
+(def reader-opts
+  {:eof eof
+   :readers (assoc graph/node-reader 'a/r re-pattern)})
 
 (defn process-args
   [args]
-  (loop [result {:query "-"} [a & [arg & rargs :as rem] :as args] args]
+  (loop [result {:interactive? false
+                 :query "-"}
+         [a & [arg & rargs :as rem] :as args] args]
     (if-not (seq args)
       result
       (let [[r more] (cond
                        (#{"-e" "-q"} a) [(assoc result :query arg) rargs]
                        (= a "-f") [(assoc result :file arg) rargs]
                        (#{"-?" "--help"} a) [(assoc result :help true) rem]
+                       (= a "--interactive") [(assoc result :interactive? true) rem]
                        (s/starts-with? a "asami:") [(assoc result :url a) rem])]
         (recur r more)))))
 
 (defn read-data-file
   [f]
-  (let [text (slurp (if (= f "-") *in* f))]
-    (if (s/ends-with? f ".json")
-      (json/parse-string text)
+  (if (s/ends-with? f ".json")
+    (json/parse-string (slurp f))
+    (let [text (slurp (if (= f "-") *in* f))]
       (edn/read-string reader-opts text))))
 
 (defn load-data-file
   [conn f]
   (:db-after @(asami/transact conn {:tx-data (read-data-file f)})))
 
-(defn read-queries
-  [q]
-  (let [text (s/trim (if (= q "-") (slurp *in*) q))
-        mtext (s/split text #";")
-        all (mapv (fn [text]
-                    (let [txt (if (= \: (first (s/trim text))) (str "[" text "]") text)]
-                      (edn/read-string reader-opts txt)))
-                  mtext)]
-    (if (= 1 (count all))
-      (let [[qv] all]
-        (if (vector? (first qv)) qv all))
-      all)))
+(defn derive-database [{:keys [file url]}]
+  (let [conn (asami/connect url)]
+    (if file
+      (load-data-file conn file)
+      (asami/db conn))))
+
+(defn derive-stream [{:keys [interactive? query]}]
+  (let [input (if (or (= query "-") interactive?)
+                *in*
+                (.getBytes ^String query))]
+    (PushbackReader. (io/reader input))))
+
+(def ^:const query-separator
+  (int \;))
+
+(defn read-queries [^PushbackReader stream]
+  (loop []
+    (let [c (.read stream)]
+      (cond
+        (or (Character/isWhitespace c) (= c query-separator))
+        (recur)
+
+        (= c -1)
```

Comment:
I'd like a `^:const` here.
Maybe `end-of-stream` since `eof` is the sentinel value?